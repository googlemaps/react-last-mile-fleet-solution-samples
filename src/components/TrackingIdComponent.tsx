/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import { View, Text, TextInput, Pressable, StyleSheet } from 'react-native';

const TaskIdComponent = ({ setTrackingId }) => {
  const [text, setText] = useState<string>();

  const onPress = () => setTrackingId(text);
  const onKeyPress = (e) => {
    if (e.nativeEvent.key === 'Enter') onPress();
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.input}
        placeholder={'Tracking ID'}
        onChangeText={(text) => setText(text)}
        onKeyPress={onKeyPress}
      />
      <Pressable style={styles.button} onPress={onPress}>
        <Text style={styles.text}>Find</Text>
      </Pressable>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    marginBottom: 10,
  },
  input: {
    marginLeft: 15,
    padding: 7,
    width: 290,
    borderWidth: 1,
    paddingLeft: 10,
    fontSize: 12,
    borderRadius: 5,
  },
  button: {
    marginLeft: 10,
    backgroundColor: '#2460ad',
    padding: 7,
    width: 50,
    borderRadius: 5,
  },
  text: {
    color: '#FFFFFF',
    fontSize: 16,
    textAlign: 'center',
  },
});

export default TaskIdComponent;
