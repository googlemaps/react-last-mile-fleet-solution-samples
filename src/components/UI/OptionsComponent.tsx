/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import { View, StyleSheet, Text, Pressable, Modal } from 'react-native';
import CheckBoxComponent from './CheckBoxComponent';

const OptionsComponent = ({ setMapOptions }) => {
  const [showAnticipatedRoutePolyline, setAnticipatedRoutePolyline] =
    useState<boolean>(true);
  const [showTakenRoutePolyline, setTakenRoutePolyline] =
    useState<boolean>(true);

  const [modalVisible, setModalVisible] = useState<boolean>(false);
  const [isOptionsChanged, setIsOptionsChanged] = useState<boolean>(false);

  const applyButtonPressHandler = () => {
    setIsOptionsChanged(false);
    setModalVisible(!modalVisible);
    setMapOptions({
      showAnticipatedRoutePolyline: showAnticipatedRoutePolyline,
      showTakenRoutePolyline: showTakenRoutePolyline,
    });
  };

  const applyButtonStyleHandler = () => {
    if (isOptionsChanged) {
      return {
        marginLeft: 10,
        backgroundColor: '#2460ad',
        padding: 5,
        width: 50,
        borderRadius: 2,
      };
    }

    return {
      marginLeft: 10,
      backgroundColor: '#A9A9A9',
      padding: 5,
      width: 50,
      borderRadius: 2,
      cursor: 'not-allowed',
    };
  };

  return (
    <View>
      <Pressable style={styles.button} onPress={() => setModalVisible(true)}>
        <Text style={styles.optionsText}>Options</Text>
      </Pressable>

      <Modal transparent={true} visible={modalVisible}>
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <View style={styles.container}>
              <Text style={styles.header}>Polyline visibility</Text>
              <View style={styles.checkboxContainer}>
                <CheckBoxComponent
                  label={'Show Anticipated Route Polyline'}
                  setRoute={setAnticipatedRoutePolyline}
                  option={showAnticipatedRoutePolyline}
                  setIsOptionsChanged={setIsOptionsChanged}
                />
                <CheckBoxComponent
                  label={'Show Taken Route Polyline'}
                  setRoute={setTakenRoutePolyline}
                  option={showTakenRoutePolyline}
                  setIsOptionsChanged={setIsOptionsChanged}
                />
              </View>
            </View>
            <View style={styles.modalButtonContainer}>
              <Pressable
                style={styles.cancelbutton}
                onPress={() => setModalVisible(!modalVisible)}
              >
                <Text style={styles.text}>Cancel</Text>
              </Pressable>
              <Pressable
                style={applyButtonStyleHandler}
                onPress={applyButtonPressHandler}
                disabled={!isOptionsChanged}
              >
                <Text style={styles.applytext}>Apply</Text>
              </Pressable>
            </View>
          </View>
        </View>
      </Modal>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    marginBottom: 20,
  },
  checkboxContainer: {
    flexDirection: 'column',
    marginLeft: 10,
  },
  header: {
    fontSize: 16,
    marginBottom: 10,
  },
  button: {
    backgroundColor: '#e8e8e8',
    borderWidth: 1,
    padding: 5,
    width: 70,
    borderRadius: 5,
  },
  applybutton: {
    marginLeft: 10,
    backgroundColor: '#2460ad',
    padding: 5,
    width: 50,
    borderRadius: 2,
  },
  cancelbutton: {
    marginLeft: 10,
    backgroundColor: '#e8e8e8',
    borderWidth: 1,
    padding: 5,
    width: 50,
    borderRadius: 2,
  },
  text: {
    fontSize: 12,
    textAlign: 'center',
  },
  applytext: {
    fontSize: 12,
    textAlign: 'center',
    color: '#FFFFFF',
    fontWeight: 'bold',
  },
  optionsText: {
    fontSize: 16,
    fontWeight: 'bold',
  },
  centeredView: {
    flex: 1,
    marginTop: 85,
    marginLeft: 70,
  },
  modalView: {
    width: 400,
    margin: 20,
    backgroundColor: '#F0F0F0',
    borderRadius: 20,
    padding: 25,
  },
  modalButtonContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
});

export default OptionsComponent;
