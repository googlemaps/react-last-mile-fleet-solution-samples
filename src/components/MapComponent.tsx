/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState, useEffect, useRef } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import TaskInformation from './TaskInformation';
import OptionsComponent from './UI/OptionsComponent';
import TrackingIdComponent from './TrackingIdComponent';
import { PROVIDER_URL, PROVIDER_PROJECT_ID, DEFAULT_POLLING_INTERVAL_MS, DEFAULT_MAP_OPTIONS } from "../utils/consts";


const MapComponent = () => {
  const ref = useRef();
  const trackingId = useRef('');
  const journeySharingMap = useRef();
  const [mapOptions, setMapOptions] = useState({});
  const [task, setTask] = useState({
    status: '',
    type: '',
    outcome: '',
    numStops: 0,
    estimatedCompletionTime: new Date(),
    journeySegments: [],
  });
  const [error, setError] = useState();

  const authTokenFetcher = async () => {
    const response = await fetch(
      `${PROVIDER_URL}/token/delivery_consumer/${trackingId.current}`
    );
    const responseJson = await response.json();

    return {
      token: responseJson.token,
      expiresInSeconds: 3300,
    };
  };

  const locationProvider = new google.maps.journeySharing.FleetEngineShipmentLocationProvider({
    projectId: PROVIDER_PROJECT_ID,
    authTokenFetcher,
    trackingId: trackingId.current,
    pollingIntervalMillis: DEFAULT_POLLING_INTERVAL_MS,
  });

  const setTrackingId = (newTrackingId) => {
    trackingId.current = newTrackingId;
    journeySharingMap.current.locationProvider.trackingId = newTrackingId;
    setError(undefined);
  }

  useEffect(() => {
    locationProvider.addListener('error', e => {
      setError(e.error.message);
    });

    locationProvider.addListener('update', e => {
      setTask((prev) => ({
        ...prev,
        status: e.task?.status,
        type: e.task?.type,
        outcome: e.task?.outcome,
        numStops: e.task?.remainingVehicleJourneySegments?.length,
        estimatedCompletionTime: e.task?.estimatedCompletionTime,
        journeySegments: e.task?.remainingVehicleJourneySegments,
      }))
    });

    journeySharingMap.current = new google.maps.journeySharing.JourneySharingMapView({
      element: ref.current,
      locationProvider,
      ...mapOptions
    });

    journeySharingMap.current.map.setOptions(DEFAULT_MAP_OPTIONS);
  }, [mapOptions]);

  return (
    <View>
      <View style={style.stack}>
        <OptionsComponent setMapOptions={setMapOptions} />
        <TrackingIdComponent setTrackingId={setTrackingId} />
        <Text style={style.heading}>Task information</Text>
        <TaskInformation style={style.infoBlock} error={error} task={task} trackingId={trackingId} />
      </View>
      <View style={style.map} ref={ref} />
    </View>
  )
}

const style = StyleSheet.create({
  map: {
    height: '87vh',
    width: '75%',
    position: 'absolute',
    right: 50,
  },
  infoBlock: {
    marginLeft: '30px',
    position: 'absolute',
    width: '25%',
    padding: '10px',
  },
  header: {
    fontSize: '2em',
    fontWeight: 'bold',
    textAlign: 'center',
    marginVertical: 20
  },
  heading: {
    fontSize: '1.6rem',
    fontWeight: 'bold',
    marginVertical: 30,
    marginLeft: 20,
  },
  stack: {
    marginLeft: 10,
    position: 'absolute',
    width: '25%',
    padding: 15,
  },
});

export default MapComponent;
