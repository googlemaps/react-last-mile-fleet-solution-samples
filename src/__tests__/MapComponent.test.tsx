/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MapComponent from '../components/MapComponent';
import TrackingIdComponent from '../components/TrackingIdComponent';
import OptionsComponent from '../components/UI/OptionsComponent';

beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
});

describe('MapComponent', () => {
    it('matches previous snapshot', async () => {
        const wrapper = shallow(<MapComponent />);
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it('renders the app correctly', () => {
        const wrapper = shallow(<MapComponent />);
        expect(wrapper.debug()).toBeTruthy();
    });

    it('calls setTrackingId on "Find" press', () => {
        const setTrackingId = jest.fn();
        const wrapper = shallow(<TrackingIdComponent setTrackingId={setTrackingId} />);
        wrapper.find('TextInput').simulate('changeText', 'Hello');
        wrapper.update();
        wrapper.find('Pressable').simulate('press');

        expect(setTrackingId).toHaveBeenCalled();
        expect(setTrackingId).toHaveBeenCalledWith('Hello');
    });

    it('renders apply button disabled', () => {
        const setMapOptions = jest.fn();
        const wrapper = shallow(<OptionsComponent setMapOptions={setMapOptions} />);
        const apply = wrapper.find('Pressable').at(2);

        expect(apply.props().disabled).toBeTruthy();
    });
});
