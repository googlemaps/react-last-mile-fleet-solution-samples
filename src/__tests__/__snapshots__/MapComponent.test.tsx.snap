// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapComponent matches previous snapshot 1`] = `
"<View>
  <TaskIdComponent setTrackingId={[Function: setTrackingId]} />
  <View style={112}>
    <View style={115}>
      <OptionsComponent setMapOptions={[Function: setMapOptions]} />
      <TaskInformation error={[undefined]} task={{...}} trackingId=\\"\\" />
    </View>
    <View style={114}>
      <View style={113} />
    </View>
  </View>
</View>"
`;
